feat: add a script to calculate the genesis rewards for testnet validators

TimmyExogenous opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Added a CLI/tool to compute time-weighted reward distributions from multiple genesis snapshots: discovers and orders snapshots, defines time periods, allocates a global reward pool by period, classifies validators as internal/external, and splits group pools accordingly.
  * Distributes rewards to validators by voting power across periods, deduplicates and aggregates validator entries, emits summaries and a detailed JSON report, and logs non-fatal data issues.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
add a script to calculate the genesis rewards for testnet validators

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/main' into genesis-validator-rewards

TimmyExogenous committed:
add some check for the genesis validator reward script

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,466 @@
+  const content = await fs.readFile(filePath, 'utf-8');
+  const json = JSON.parse(content);
+  
+  const genesisTimeStr = json.genesis_time;
+  if (!genesisTimeStr) {
+    throw new Error(`Missing 'genesis_time' in ${path.basename(filePath)}`);
+  }
+  const genesisTime = new Date(genesisTimeStr).getTime();
```
It will be added in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,466 @@
+    valSet = [];
+  }
+
+  const operatorRecords = json.app_state?.operator?.operator_records || [];
+  const consensusKeyToOpAddr = {};
+  operatorRecords.forEach(record => {
+    if (record.chains) {
+      record.chains.forEach(chainInfo => {
+        consensusKeyToOpAddr[chainInfo.consensus_key.toLowerCase()] = record.operator_address;
+      });
+    }
+  });
```
It will be added in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,466 @@
+    const [genesisDir, configDir, outputDir] = args;
+
+    const configPath = path.join(configDir, 'config.json');
+    const configRaw = await fs.readFile(configPath, 'utf-8');
+    const config = JSON.parse(configRaw);
+
+    let endTimeMs = Date.now();
+    if (config.reward_end_time) {
+      endTimeMs = new Date(config.reward_end_time).getTime();
+    } else {
+      console.warn("[Warn] 'reward_end_time' not found in config. Using current time.");
+    }
```
It will be added in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,466 @@
+
+      const internalVals = [];
+      const externalVals = [];
+      let internalPower = new BigNumber(0);
+      let externalPower = new BigNumber(0);
+      let periodTotalPower = new BigNumber(0);
+
+      for (const val of period.valSet) {
+        const pubKey = val.public_key.toLowerCase();
+        const power = new BigNumber(val.power);
```
It will be added in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,466 @@
+    await fs.writeFile(finalPath, JSON.stringify(outputData, null, 2));
+
+    console.log(`Success! Result written to: ${finalPath}`);
+
+  } catch (error) {
+    console.error('Error:', error.message, error.stack);
+  }
```
It will be added in the next commit.

bwhour commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,479 @@
+      const nextTime = (i === parsedFiles.length - 1) ? endTimeMs : parsedFiles[i + 1].genesisTime;
+
+      const duration = nextTime - current.genesisTime;
+      if (duration <= 0) continue;
```
add some log here. maybe genesis file timestamps exceptions

TimmyExogenous replied:
Thanks, log will be added in the next commit.

cloud8little committed:
Merge branch 'main' into genesis-validator-rewards

cloud8little commented:
Please add integration of validator bech32 address before rebranding and after rebranding, for example, operator1 `exo1lzf5247us2xwhg29mutpx5gyq4aemwqgc08ev7 `and validator1 `im1lzf5247us2xwhg29mutpx5gyq4aemwqgg3ypwm `are the same validator, the reward amount should add up for them.

TimmyExogenous committed:
Merge branch 'main' of TimmyExogenous-github.com:imua-xyz/imua-contracts into genesis-validator-rewards

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,479 @@
+ * 1. Dependencies: Ensure `bignumber.js` is installed.
+ * $ npm install bignumber.js
+ *
+ * 2. Command:
+ * $ node calculate_rewards.mjs <genesis_dir> <config_dir> <output_dir>
+ * * Example:
+ * $ node calculate_rewards.mjs ./genesis_files ./config ./output
+ *
```
Thanks, it will be updated in the next commit.

TimmyExogenous committed:
address the old operator addresses before rebranding and improve logging in genesis rewards script

TimmyExogenous committed:
Merge branch 'genesis-validator-rewards' of TimmyExogenous-github.com:TimmyExogenous/imua-contracts into genesis-validator-rewards

TimmyExogenous commented:
> Please add integration of validator bech32 address before rebranding and after rebranding, for example, operator1 `exo1lzf5247us2xwhg29mutpx5gyq4aemwqgc08ev7 `and validator1 `im1lzf5247us2xwhg29mutpx5gyq4aemwqgg3ypwm `are the same validator, the reward amount should add up for them.

fixed

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,545 @@
+ * }
+ */
+
+import fs from 'fs/promises';
+import path from 'path';
+import { BigNumber } from 'bignumber.js';
+import { decode, encode } from 'bech32';
```
Thanks, it will be updated in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,545 @@
+    }
+  });
+
+  const operators = json.app_state?.operator?.operators || [];
+  const opAddrToName = {};
+  operators.forEach(op => {
+    const earningsAddr = op.earnings_addr || op.operator_info?.earnings_addr;
+    const metaInfo = op.operator_meta_info || op.operator_info?.operator_meta_info;
+    if (earningsAddr) {
+      const normalizedAddr = normalizeAddress(earningsAddr);
+      opAddrToName[normalizedAddr] = metaInfo || "";
+    }
+  });
```
The `generate.mjs` has been updated to be compatible with the new genesis structure. However, the reward calculation script is currently used to process the historical testnet genesis file (which uses the old structure). This is fine for now; I will update the script once we have a genesis file using the new structure.

TimmyExogenous committed:
refine the code comment in genesis_validator_rewards.mjs

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,545 @@
+  try {
+    const args = process.argv.slice(2);
+    if (args.length < 3) {
+      console.error("Usage: node calculate_rewards.mjs <genesis_dir> <config_dir> <output_dir>");
```
Thanks, it will be updated in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,545 @@
+    const totalDurationDays = new BigNumber(totalDurationMs).div(1000 * 60 * 60 * 24);
+    console.log(`Total Duration: ${totalDurationDays.toFixed(2)} days.`);
+
+    const totalSupply = new BigNumber(config.total_supply);
+    const totalGlobalReward = totalSupply.times(config.genesis_validator_ratio);
```
Thanks, the check will be added in the next commit.

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,545 @@
+      const hasInternal = internalVals.length > 0;
+      const hasExternal = externalVals.length > 0;
+
+      let internalPool = new BigNumber(0);
+      let externalPool = new BigNumber(0);
+
+      if (config.distinguish_internal_external) {
+        // Handle Empty Groups with Fixed Ratio
+        if (hasInternal && hasExternal) {
+          // Normal case: Both groups exist, apply fixed ratio
+          const extRatio = new BigNumber(config.external_validator_ratio);
+          externalPool = periodPool.times(extRatio);
+          internalPool = periodPool.minus(externalPool);
```
Thanks, the check will be added in the next commit.

TimmyExogenous committed:
add some config validations in genesis_validator_rewards.mjs

TimmyExogenous committed:
add zero-value check for totalSupply and totalGlobalReward in genesis_validator_rewards.mjs

TimmyExogenous commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,554 @@
+    // Prepare & Enrich Results
+    // ------------------------------------------------------------------
+
+    const allResults = Array.from(globalValidatorMap.values()).map(v => ({
+      operator_address: v.operator_address,
+      name: v.name,
+      type: v.type,
+      periods_participated: v.periods_active,
+      reward_amount: v.total_reward.toFixed(18),
+      percentage_of_total_supply: v.total_reward.div(totalSupply).toString(),
+      percentage_of_total_reward: v.total_reward.div(totalGlobalReward).toString()
+    }));
```
Thanks, I will add a zero-value check in the next commit.

cloud8little commented:
> Please add integration of validator bech32 address before rebranding and after rebranding, for example, operator1 `exo1lzf5247us2xwhg29mutpx5gyq4aemwqgc08ev7 `and validator1 `im1lzf5247us2xwhg29mutpx5gyq4aemwqgg3ypwm `are the same validator, the reward amount should add up for them.

Verified. the amount for same validator but different address after rebranding has been merged.

bwhour commented on script/bootstrap/genesis_validator_rewards.mjs:
```diff
@@ -0,0 +1,554 @@
+        const power = new BigNumber(val.power);
+
+        // Filter out validators with zero voting power to prevent reward leaks.
+        if (power.lte(0)) {
```
prevent possible dirty data 'NaN/Infinity'
> if (!power.isFinite() || power.isNaN() || power.lte(0)) {
>   console.warn(`[Warn] Invalid validator power in ${period.fileName}: ${val.power}`);
>   continue;
> }

TimmyExogenous replied:
Fixed. Added an explicit guard for invalid power values (!power.isFinite() || power.isNaN() || power.lte(0)) and now log a warning before skipping the validator entry.

TimmyExogenous committed:
add NaN/Infinity guard for validator power in genesis rewards script

Skip validator entries with non-finite power values and log a warning to
prevent dirty input from affecting reward calculations.

Made-with: Cursor
